File: app/src/main/java/net/itsjustsomedude/tokens/db/Converters.java
package net.itsjustsomedude.tokens.db;

import androidx.room.TypeConverter;

import java.util.Calendar;

public class Converters {
	@TypeConverter
	public static Long fromCalendarToLong(Calendar calendar) {
		return calendar == null ? null : calendar.getTimeInMillis();
	}

	// Convert Long to Calendar
	@TypeConverter
	public static Calendar fromLongToCalendar(Long timestamp) {
		if (timestamp == null) {
			return null;
		}
		Calendar calendar = Calendar.getInstance();
		calendar.setTimeInMillis(timestamp);
		return calendar;
	}

	// Convert Integer to Boolean
	@TypeConverter
	public static Boolean fromIntegerToBoolean(Integer value) {
		return value == null ? null : value != 0;
	}

	// Convert Boolean to Integer
	@TypeConverter
	public static Integer fromBooleanToInteger(Boolean value) {
		return value == null ? null : value ? 1 : 0;
	}
}

File: app/src/main/java/net/itsjustsomedude/tokens/db/Converters.kt
package net.itsjustsomedude.tokens.db

import androidx.room.TypeConverter
import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.util.Calendar

object Converters {
    @TypeConverter
    fun fromCalendarToLong(calendar: Calendar?): Long? {
        return calendar?.timeInMillis
    }

    @TypeConverter
    fun fromLongToCalendar(timestamp: Long?): Calendar? {
        if (timestamp == null)
            return null

        return Calendar.getInstance().apply {
            timeInMillis = timestamp
        }
    }

    @TypeConverter
    fun fromIntegerToBoolean(value: Int?): Boolean? {
        return if (value == null) null else value != 0
    }

    @TypeConverter
    fun fromBooleanToInteger(value: Boolean?): Int? {
        return if (value == null) null else if (value) 1 else 0
    }

    @ExperimentalSerializationApi
    @TypeConverter
    fun fromStringList(value: List<String>?): String? {
        return value?.let { Json.encodeToString(it) }
    }

    @ExperimentalSerializationApi
    @TypeConverter
    fun toStringList(value: String?): List<String>? {
        return value?.let { Json.decodeFromString<List<String>>(it) }
    }

    @ExperimentalSerializationApi
    @TypeConverter
    fun toStringIntMap(value: Map<String, Int>?): String? {
        return value?.let { Json.encodeToString(it) }
    }

    @ExperimentalSerializationApi
    @TypeConverter
    fun fromStringIntMap(value: String?): Map<String, Int>? {
        return value?.let { Json.decodeFromString(it) }
    }
}
